File: ClassEval_T/java/solutuon/TriCalculator.java
package org.example;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;

public class TriCalculator {

    public TriCalculator() {
    }

    public double cos(double x) {
        return round(taylor(x, 50), 10);
    }

    public BigDecimal factorial(int a) {
        BigDecimal result = BigDecimal.ONE;
        while (a > 1) {
            result = result.multiply(BigDecimal.valueOf(a));
            a--;
        }
        return result;
    }

    public double taylor(double x, int n) {
        BigDecimal a = BigDecimal.ONE;
        x = x / 180 * Math.PI;
        int count = 1;
        for (int k = 1; k < n; k++) {
            BigDecimal term = BigDecimal.valueOf(Math.pow(x, 2 * k))
                    .divide(factorial(2 * k), MathContext.DECIMAL128);
            if (count % 2 != 0) {
                a = a.subtract(term);
            } else {
                a = a.add(term);
            }
            count++;
        }
        return a.doubleValue();
    }

    public double sin(double x) {
        x = x / 180 * Math.PI;
        double g = 0;
        double t = x;
        int n = 1;

        while (Math.abs(t) >= 1e-15) {
            g += t;
            n += 1;
            t = -t * x * x / (2 * n - 1) / (2 * n - 2);
        }
        return round(g, 10);
    }

    public Double tan(double x) {
        double cosValue = cos(x);
        if (cosValue != 0) {
            return round(sin(x) / cosValue, 10);
        } else {
            return Double.NaN;
        }
    }

    private double round(double value, int places) {
        BigDecimal bd = BigDecimal.valueOf(value);
        bd = bd.setScale(places, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }
}

File: ClassEval_T/cpp/solution/TriCalculator.cpp

#include <cmath>
#include <limits>
#include <iostream>

class TriCalculator{
public:
    TriCalculator();
    double cos(double x);
    int factorial(int a);
    double taylor(double x, int n);
    double sin(double x);
    double tan(double x);

private:
    static constexpr double PI = 3.14159265358979323846;
    double round(double value, int precision);
};


double TriCalculator::round(double value, int precision) {
    double factor = std::pow(10.0, precision);
    return std::round(value * factor) / factor;
}

TriCalculator::TriCalculator() {
    
}

double TriCalculator::cos(double x) {
    double result = taylor(x, 17);
    std::cout << "Taylor result for cos: " << result << std::endl;
    return round(result, 10);
}

int TriCalculator::factorial(int a) {
    if (a < 0) return 0; 
    if (a == 0) return 1; 
    int b = 1;
    for (int i = 1; i <= a; ++i) {
        b *= i;
    }
    return b;
}

double TriCalculator::taylor(double x, int n) {
    double a = 0.0; 
    x = x / 180.0 * PI; 
    int sign = 1; 

    for (int k = 0; k < n; ++k) {
        double term = std::pow(x, 2 * k) / factorial(2 * k);
        if (k % 2 == 0) {
            a += term;
        }
        else {
            a -= term;
        }
    }
    return a;
}

double TriCalculator::sin(double x) {
    x = x / 180.0 * PI;
    double g = 0.0;
    double t = x;
    int n = 1;

    while (std::fabs(t) >= 1e-15) {
        g += t;
        ++n;
        t = -t * x * x / ((2 * n - 1) * (2 * n - 2));
    }
    return round(g, 10);
}

double TriCalculator::tan(double x) {
    double cosine = cos(x);
    if (std::fabs(cosine) > 0.01) {
        
        double result = sin(x) / cosine;
        
        return round(result, 10);
    }
    else {
        return std::numeric_limits<double>::quiet_NaN(); 
    }
}
